### PR TITLE
Use RTLD_NODELETE=true when loading libraries

### LINKED ISSUE #806: Components registered by multiple systems cause segfaults on macOS
## Environment
* OS Version: macOS Mojave
* Source or binary build? source, ign-gazebo3

## Description
* Expected behavior: No segfault
* Actual behavior:  If a component's header file is included by a system, but not by the translation units that comprise `libignition-gazebo.so`, the component is registered only when that system is loaded. Having two such systems can cause segfaults on macOS. This is manifested in number of flaky tests documented in #444 and other issues. From what I can tell, the problem arises when a test with two or more test cases that create a server and load worlds with those systems, that is

```c++

TEST(Foo, TEST1)
{
  Server server(...); // Load world with two or more systems that include a component's header
  server.Run(...);
  ...
}
 
TEST(Foo, TEST2)
{
  Server server(...);  // Load world with two or more systems that include a component's header
  server.Run(...);
  ...
}
```

An example is test/integration/joint_state_publisher_system.cc

## Steps to reproduce

1. Build ign-gazebo with Address Sanitizer
2. Run `INTEGRATION_joint_state_publisher_system`

## Output

Address sanitizer output from running `INTEGRATION_joint_state_publisher_system`

```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==89646==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x00010338e203 bp 0x7ffeeccaa870 sp 0x7ffeeccaa6c0 T0)
==89646==The signal is caused by a READ memory access.
==89646==Hint: address points to the zero page.
    #0 0x10338e202 in ignition::gazebo::v3::components::Factory::NewStorage(unsigned long long const&) Factory.hh:259
    #1 0x10338b085 in ignition::gazebo::EntityComponentManagerPrivate::CreateComponentStorage(unsigned long long) EntityComponentManager.cc:622
    #2 0x10338a3aa in ignition::gazebo::v3::EntityComponentManager::CreateComponentImplementation(unsigned long long, unsigned long long, ignition::gazebo::v3::components::BaseComponent const*) EntityComponentManager.cc:483
    #3 0x11a3c82e6 in std::__1::pair<unsigned long long, int> ignition::gazebo::v3::EntityComponentManager::CreateComponent<ignition::gazebo::v3::components::Component<std::__1::vector<double, std::__1::allocator<double> >, ignition::gazebo::v3::components::JointPositionTag, ignition::gazebo::v3::serializers::VectorDoubleSerializer> >(unsigned long long, ignition::gazebo::v3::components::Component<std::__1::vector<double, std::__1::allocator<double> >, ignition::gazebo::v3::components::JointPositionTag, ignition::gazebo::v3::serializers::VectorDoubleSerializer> const&) EntityComponentManager.hh:86
    #4 0x11a3c5f68 in ignition::gazebo::v3::systems::JointStatePublisher::CreateComponents(ignition::gazebo::v3::EntityComponentManager&, unsigned long long) JointStatePublisher.cc:110
    #5 0x11a3c4b26 in ignition::gazebo::v3::systems::JointStatePublisher::Configure(unsigned long long const&, std::__1::shared_ptr<sdf::v9::Element const> const&, ignition::gazebo::v3::EntityComponentManager&, ignition::gazebo::v3::EventManager&) JointStatePublisher.cc:71
    #6 0x103acf559 in ignition::gazebo::v3::SimulationRunner::LoadPlugin(unsigned long long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<sdf::v9::Element> const&) SimulationRunner.cc:769
    #7 0x103aba5d1 in ignition::gazebo::v3::SimulationRunner::LoadPlugins(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&) SimulationRunner.cc:903
    #8 0x103b61e2e in decltype(*(std::__1::forward<ignition::gazebo::v3::SimulationRunner*&>(fp0)).*fp(std::__1::forward<unsigned long long>(fp1), std::__1::forward<std::__1::shared_ptr<sdf::v9::Element> >(fp1))) std::__1::__invoke<void (ignition::gazebo::v3::SimulationRunner::*&)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*&, unsigned long long, std::__1::shared_ptr<sdf::v9::Element>, void>(void (ignition::gazebo::v3::SimulationRunner::*&&&)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*&&&, unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&) type_traits:4302
    #9 0x103b61c59 in std::__1::__bind_return<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), std::__1::tuple<ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1>, std::__1::placeholders::__ph<2> >, std::__1::tuple<unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&>, __is_valid_bind_return<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), std::__1::tuple<ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1>, std::__1::placeholders::__ph<2> >, std::__1::tuple<unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&> >::value>::type std::__1::__apply_functor<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), std::__1::tuple<ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1>, std::__1::placeholders::__ph<2> >, 0ul, 1ul, 2ul, std::__1::tuple<unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&> >(void (ignition::gazebo::v3::SimulationRunner::*&)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), std::__1::tuple<ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1>, std::__1::placeholders::__ph<2> >&, std::__1::__tuple_indices<0ul, 1ul, 2ul>, std::__1::tuple<unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&>&&) functional:2644
    #10 0x103b61b64 in std::__1::__bind_return<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), std::__1::tuple<ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1>, std::__1::placeholders::__ph<2> >, std::__1::tuple<unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&>, __is_valid_bind_return<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), std::__1::tuple<ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1>, std::__1::placeholders::__ph<2> >, std::__1::tuple<unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&> >::value>::type std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&>::operator()<unsigned long long, std::__1::shared_ptr<sdf::v9::Element> >(unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&) functional:2677
    #11 0x103b619f6 in decltype(std::__1::forward<std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&>&>(fp)(std::__1::forward<unsigned long long>(fp0), std::__1::forward<std::__1::shared_ptr<sdf::v9::Element> >(fp0))) std::__1::__invoke<std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&>&, unsigned long long, std::__1::shared_ptr<sdf::v9::Element> >(std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&>&&&, unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&) type_traits:4361
    #12 0x103b61986 in void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&>&, unsigned long long, std::__1::shared_ptr<sdf::v9::Element> >(std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&>&&&, unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&) __functional_base:349
    #13 0x103b61916 in std::__1::__function::__alloc_func<std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&>, std::__1::allocator<std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&> >, void (unsigned long long, std::__1::shared_ptr<sdf::v9::Element>)>::operator()(unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&) functional:1527
    #14 0x103b5e055 in std::__1::__function::__func<std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&>, std::__1::allocator<std::__1::__bind<void (ignition::gazebo::v3::SimulationRunner::*)(unsigned long long, std::__1::shared_ptr<sdf::v9::Element> const&), ignition::gazebo::v3::SimulationRunner*, std::__1::placeholders::__ph<1> const&, std::__1::placeholders::__ph<2> const&> >, void (unsigned long long, std::__1::shared_ptr<sdf::v9::Element>)>::operator()(unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&) functional:1651
    #15 0x10360b0f8 in std::__1::__function::__value_func<void (unsigned long long, std::__1::shared_ptr<sdf::v9::Element>)>::operator()(unsigned long long&&, std::__1::shared_ptr<sdf::v9::Element>&&) const functional:1799
    #16 0x103609839 in std::__1::function<void (unsigned long long, std::__1::shared_ptr<sdf::v9::Element>)>::operator()(unsigned long long, std::__1::shared_ptr<sdf::v9::Element>) const functional:2347
    #17 0x10393f24d in void ignition::common::EventT<void (unsigned long long, std::__1::shared_ptr<sdf::v9::Element>), ignition::gazebo::v3::events::LoadPluginsTag>::Signal<unsigned long long const&, std::__1::shared_ptr<sdf::v9::Element> const&>(unsigned long long const&&&, std::__1::shared_ptr<sdf::v9::Element> const&&&) Event.hh:146
    #18 0x1037aea6e in void ignition::gazebo::v3::EventManager::Emit<ignition::common::EventT<void (unsigned long long, std::__1::shared_ptr<sdf::v9::Element>), ignition::gazebo::v3::events::LoadPluginsTag>, unsigned long long const&, std::__1::shared_ptr<sdf::v9::Element> const&>(unsigned long long const&&&, std::__1::shared_ptr<sdf::v9::Element> const&&&) EventManager.hh:108
    #19 0x1037aa59d in ignition::gazebo::v3::SdfEntityCreator::CreateEntities(sdf::v9::Model const*) SdfEntityCreator.cc:235
    #20 0x10349f712 in ignition::gazebo::v3::LevelManager::LoadActiveEntities(std::__1::set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) LevelManager.cc:756
    #21 0x10349a06a in ignition::gazebo::v3::LevelManager::UpdateLevelsState() LevelManager.cc:709
    #22 0x103ab639f in ignition::gazebo::v3::SimulationRunner::SimulationRunner(sdf::v9::World const*, std::__1::shared_ptr<ignition::gazebo::v3::SystemLoader> const&, ignition::gazebo::v3::ServerConfig const&) SimulationRunner.cc:158
    #23 0x103ac005c in ignition::gazebo::v3::SimulationRunner::SimulationRunner(sdf::v9::World const*, std::__1::shared_ptr<ignition::gazebo::v3::SystemLoader> const&, ignition::gazebo::v3::ServerConfig const&) SimulationRunner.cc:50
    #24 0x103a4b390 in std::__1::__unique_if<ignition::gazebo::v3::SimulationRunner>::__unique_single std::__1::make_unique<ignition::gazebo::v3::SimulationRunner, sdf::v9::World const*&, std::__1::shared_ptr<ignition::gazebo::v3::SystemLoader>&, ignition::gazebo::v3::ServerConfig&>(sdf::v9::World const*&&&, std::__1::shared_ptr<ignition::gazebo::v3::SystemLoader>&&&, ignition::gazebo::v3::ServerConfig&&&) memory:3132
    #25 0x103a4acda in ignition::gazebo::v3::ServerPrivate::CreateEntities() ServerPrivate.cc:364
    #26 0x1039d25d0 in ignition::gazebo::v3::Server::Server(ignition::gazebo::v3::ServerConfig const&) Server.cc:191
    #27 0x1039d41dc in ignition::gazebo::v3::Server::Server(ignition::gazebo::v3::ServerConfig const&) Server.cc:80
    #28 0x102f59fff in JointStatePublisherTest_LimitedPublisher_Test::TestBody() joint_state_publisher_system.cc:100
    #29 0x10307dbf2 in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) gtest.cc:2437
    #30 0x102ff7691 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) gtest.cc:2473
    #31 0x102ff715c in testing::Test::Run() gtest.cc:2512
    #32 0x102ffa049 in testing::TestInfo::Run() gtest.cc:2688
    #33 0x102ffcd6d in testing::TestSuite::Run() gtest.cc:2820
    #34 0x10301c500 in testing::internal::UnitTestImpl::RunAllTests() gtest.cc:5342
    #35 0x10308b882 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) gtest.cc:2437
    #36 0x10301b400 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) gtest.cc:2473
    #37 0x10301ae98 in testing::UnitTest::Run() gtest.cc:4929
    #38 0x1030a3720 in RUN_ALL_TESTS() gtest.h:2473
    #39 0x1030a3695 in main gtest_main.cc:45
    #40 0x7fff697b93d4 in start (libdyld.dylib:x86_64+0x163d4)

==89646==Register values:
rax = 0x0000000113b6d810  rbx = 0x00007ffeeccaa740  rcx = 0x696f4a636974616d  rdx = 0x000010002276db02
rdi = 0x00007ffeeccaa720  rsi = 0x00006020000bbb10  rbp = 0x00007ffeeccaa870  rsp = 0x00007ffeeccaa6c0
 r8 = 0x000000010402ba30   r9 = 0x000000010402ba30  r10 = 0x05c5fa1df8df0041  r11 = 0x00000000000001c0
r12 = 0x0000100000000000  r13 = 0x00001fffdd995568  r14 = 0x00001fffdd995514  r15 = 0x00001fffdd995560
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV Factory.hh:259 in ignition::gazebo::v3::components::Factory::NewStorage(unsigned long long const&)
==89646==ABORTING
```

### PR DESCRIPTION
**Pull request by azeey**
Fixes (Potentially): https://github.com/gazebosim/gz-sim/issues/806

## Summary

Gazebo Garden [crashes](https://github.com/gazebosim/gz-sim/issues/1443) at exit because the destructor for an `EventT` object created by a gui plugin is accesses after the plugin is unloaded. For that particular crash, the plugin in question is in gz-gui, so I've submitted a fix there (https://github.com/gazebosim/gz-gui/pull/469). This PR is proactively using `RTLD_NODELETE` to prevent future bugs of this nature. 

I believe this might also fix #806 because if the shared libraries never get deleted, when a plugin goes through a load--unload--load cycle, the components (more specifically, the vtable of the components) it registers will be located in the same memory address. This was not the case before, which I think was the culprit for many of the flakiness in macOS.

More context:
* https://github.com/gazebosim/gz-common/pull/367

Requires:
*  https://github.com/gazebosim/gz-plugin/pull/102

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by nkoenig — APPROVED** [review]

**Comment by chapulina** [comment]
@osrf-jenkins run tests please

File: src/gui/Gui_clean_exit_TEST.cc [hunk 1]
```diff
@@ -77,9 +77,7 @@ void startBoth(const std::string &_fileName)
 }
 
 /////////////////////////////////////////////////
-/// TODO (azeey) Temporarliy disabled until
-/// https://github.com/gazebosim/gz-sim/issues/1443 is resolved
-TEST_P(GazeboDeathTest, DISABLED_CleanExit)
+TEST_P(GazeboDeathTest, CleanExit)
```
Comment (chapulina) [inline src/gui/Gui_clean_exit_TEST.cc:80]: This test is failing on all platforms 😕

Comment (chapulina) [reply src/gui/Gui_clean_exit_TEST.cc:80]: Actually, this is using nightlies of `plugin2`, but not of `gui7` yet, let me trigger some:

* [Build Status](https://build.osrfoundation.org/job/gz-gui7-debbuilder/614/)
* [Build Status](https://build.osrfoundation.org/job/gz-gui7-debbuilder/615/)

Comment (chapulina) [reply src/gui/Gui_clean_exit_TEST.cc:80]: Ok, now it passes on Ubuntu, but it's failing on macOS:

```
61: [ RUN      ] WorldFiles/GazeboDeathTest.CleanExit/0
61: 
61: [WARNING] /Users/jenkins/workspace/ignition_gazebo-ci-pr_any-homebrew-amd64/ign-gazebo/test/gtest_vendor/src/gtest-death-test.cc:1108:: Death tests use fork(), which is unsafe particularly in a threaded context. For this test, Google Test detected 3 threads. See https://github.com/google/googletest/blob/master/docs/advanced.md#death-tests-and-threads for more explanation and suggested solutions, especially if this is the last message you see before your test times out.
61: /Users/jenkins/workspace/ignition_gazebo-ci-pr_any-homebrew-amd64/ign-gazebo/src/gui/Gui_clean_exit_TEST.cc:97: Failure
61: Death test: startBoth(sdfFile)
61:     Result: died but not with expected exit code:
61:             Terminated by signal 11
61: Actual msg:
61: [  DEATH   ] The process has forked and you cannot use this CoreFoundation functionality safely. You MUST exec().
61: [  DEATH   ] Break on __THE_PROCESS_HAS_FORKED_AND_YOU_CANNOT_USE_THIS_COREFOUNDATION_FUNCTIONALITY___YOU_MUST_EXEC__() to debug.
61: [  DEATH   ] 
61: [  FAILED  ] WorldFiles/GazeboDeathTest.CleanExit/0, where GetParam() = "empty.sdf" (50 ms)
```

Windows is also failing, but it doesn't even find the shared libraries.

I'm going to disable the test for macOS and Windows.

Comment (azeey) [reply src/gui/Gui_clean_exit_TEST.cc]: Also only enabled on Linux in Fortress, so thanks for making the change in https://github.com/gazebosim/gz-sim/pull/1649/commits/f0c2c27067a9898edbad12579a8993de41faa91f

**Review by chapulina — APPROVED** [review]

**Pull request merged**